File: fotoapparat-adapters/rxjava2/src/main/java/io/fotoapparat/result/adapter/rxjava2/ObservableAdapter.java
package io.fotoapparat.result.adapter.rxjava2;

import java.util.concurrent.Future;

import io.fotoapparat.result.adapter.Adapter;
import io.reactivex.Observable;

/**
 * Adapter for {@link Observable}.
 */
public class ObservableAdapter<T> implements Adapter<T, Observable<T>> {

    private ObservableAdapter() {
    }

    /**
     * @return {@link Adapter} which adapts result to {@link Observable}.
     */
    public static <R> ObservableAdapter<R> toObservable() {
        return new ObservableAdapter<>();
    }

    @Override
    public Observable<T> adapt(Future<T> future) {
        return Observable.fromFuture(future);
    }
}

File: fotoapparat-adapters/rxjava2/src/main/java/io/fotoapparat/result/adapter/rxjava2/ObservableAdapter.kt
package io.fotoapparat.result.adapter.rxjava2

import android.annotation.SuppressLint
import io.fotoapparat.result.PendingResult
import io.reactivex.Observable
import java.util.concurrent.Future

/**
 * Adapter for [Observable].
 */
object ObservableAdapter {

    /**
     * @return Adapter which adapts result to [Observable].
     */
    @JvmStatic
    @SuppressLint("CheckResult")
    fun <T> toObservable(): Function1<Future<T>, Observable<T>> {
        return { future -> Observable.fromFuture(future) }
    }

}

/**
 * @return A [Observable] from the given [PendingResult].
 */
@SuppressLint("CheckResult")
fun <T> PendingResult<T>.toObservable(): Observable<T> {
    return adapt { future -> Observable.fromFuture(future) }
}
